internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,117 @@
+		ChunkSize = int64(sizeInMB * MiB)
+	}
+	if b.currentBuffer == nil {
+		b.currentBuffer = make([]byte, 0, ChunkSize)
+		b.currentBufferEndOffset = ChunkSize
+	}
+	if b.flushedBuffer == nil {
+		b.flushedBuffer = make([]byte, 0, ChunkSize)
+	}
+}
+
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, receivedContentStartOffset int64) error {
+	dataSize := int64(len(content))
+	receivedContentEndOffset := receivedContentStartOffset + dataSize
+
+	// If receivedContentStartOffset is not in the range of offsets stored in current
+	// buffer, trigger temp file flow for writes.
+	if receivedContentStartOffset < b.currentBufferStartOffset ||
+		b.currentBufferEndOffset < receivedContentStartOffset {
+		// TODO: finalise write and trigger temp file flow.
+		return fmt.Errorf(NonSequentialWriteError)
+	}
+
+	// If data can be completely written to the current buffer block, write it.
+	if receivedContentEndOffset <= b.currentBufferEndOffset {
+		return b.copyDataToBuffer(receivedContentStartOffset, receivedContentEndOffset, content)
+	} else {
+		// Below logic is written with the assumption that the data received in a
+		// single write call from the kernel will never exceed beyond buffer block size.
+		// This should always hold true because kernel writes never exceed 1 MiB size.
+
+		// Write the chunk of data that can be written to the current buffer block.
+		l := b.currentBufferEndOffset - receivedContentStartOffset
+		if err := b.copyDataToBuffer(receivedContentStartOffset, b.currentBufferEndOffset, content[:l]); err != nil {
+			return err
+		}
+
+		//TODO: get status of last write to GCS.
+		// If unsuccessful, trigger temp file flow.
+		// Else:
+		b.swapBuffers()
+		// TODO: trigger async upload to GCS for flushedBuffer.
+
+		// Update buffer offsets for next buffer block.
+		b.currentBufferStartOffset = b.currentBufferEndOffset
+		b.currentBufferEndOffset += ChunkSize
+		// Write the remaining data to currentBuffer.
+		return b.copyDataToBuffer(b.currentBufferStartOffset, receivedContentEndOffset, content[l:])
 	}
 }
 
-func (b *InMemoryWriteBuffer) WriteAt(data []byte, offset int64) error {
+func (b *InMemoryWriteBuffer) copyDataToBuffer(contentStartOffset, contentEndOffset int64, content []byte) error {
+	dataSize := int64(len(content))
+	si := contentStartOffset % ChunkSize
+	ei := si + dataSize
+
+	n := copy(b.currentBuffer[si:ei], content)
+	if int64(n) != dataSize {
+		return fmt.Errorf(DataNotWrittenCompletelyError, dataSize, n)
+	}
+	b.updateFileSize(contentEndOffset)
 	return nil
 }
+
+func (b *InMemoryWriteBuffer) swapBuffers() {
+	b.flushedBufferStartOffset = b.currentBufferStartOffset
+	b.flushedBufferEndOffset = b.currentBufferEndOffset
+	b.flushedBuffer, b.currentBuffer = b.currentBuffer, b.flushedBuffer
+	clear(b.currentBuffer)
+}
+
+// After successful copy of data to buffer, increments the bytes written so far.
+func (b *InMemoryWriteBuffer) updateFileSize(endOffset int64) {
+	if endOffset > b.fileSize {
```

Comment:
Will this check ever fail ? If yes, in what case ?

Reply:
Yes, if a user overwrites some previously written content which is currently in buffer.